File: languages/haskell/flipper/src/Flipper/Internal/Button.hs
{-|
Module      : Flipper.Internal.Button
Description : Internal Button Module
Copyright   : George Morgan, Travis Whitaker 2016
License     : All rights reserved.
Maintainer  : travis@flipper.io
Stability   : Provisional
Portability : Windows, POSIX

-}

module Flipper.Internal.Button (
    configure
  , read
  ) where

import Prelude hiding (read)

import Data.Word

import Flipper.Internal.Utils

configure :: IO Bool
configure = retSuc <$> c_button_configure

read :: IO Bool
read = retSuc <$> c_button_read

foreign import ccall safe "flipper/button/button.h button_configure"
    c_button_configure :: IO Word32

foreign import ccall safe "flipper/button/button.h button_read"
    c_button_read :: IO Word8

File: carbon/atmegau2/button.c
#include <flipper/button.h>


int button_configure(void) {
	BUTTON_DDR &= ~(1 << BUTTON_PIN);
	return lf_success;
}

uint8_t button_read(void) {
	return ((BUTTON_IN & (1 << BUTTON_PIN)) >> BUTTON_PIN);
}
